### PR TITLE
Feature/wsd scheduler

### PR DESCRIPTION
**Pull request by hawkoli1987**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?
Implements the Weight Stable Decay (WSD) scheduler with both linear (https://arxiv.org/pdf/2404.06395) and one-minus-square-root decay options (https://arxiv.org/html/2408.11029) as described in recent research papers. This scheduler introduces a three-phase learning rate schedule: warmup, hold, and decay, which has been shown to improve training stability and performance for large language models.

**Collection**: [Note which collection this PR will affect] 
Core

# Changelog
- Added WarmupHoldAnnealOneMinusSquareRoot scheduler that implements the 1-sqrt(t) decay pattern
- Added WarmupHoldAnnealLinear scheduler that implements the linear decay pattern
- Added WarmupHoldAnnealScheduler Lightning wrapper class that supports both decay schedules
- Added unit tests for both scheduler implementations
- Updated scheduler registry to include the new schedulers

# Usage

- You can potentially add a usage example below

```python
# Using the Lightning wrapper
from nemo.core.optim.lr_scheduler import WarmupHoldAnneal

# One-minus-square-root decay (recommended in the paper)
scheduler = WarmupHoldAnneal(
    warmup_ratio=0.01,    # 1% of steps for warmup
    hold_ratio=0.19,      # 19% of steps for hold phase
    max_steps=10000,      # Total training steps
    decay_schedule="one_minus_square_root",  # Use 1-sqrt(t) decay
    min_lr=1e-6           # Minimum learning rate
)

# Linear decay option
scheduler = WarmupHoldAnneal(
    warmup_ratio=0.01,
    hold_ratio=0.19,
    max_steps=10000,
    decay_schedule="linear",  # Use linear decay
    min_lr=1e-6
)

# Direct usage with PyTorch optimizer
from nemo.core.optim.lr_scheduler import WarmupHoldAnnealOneMinusSquareRoot

optimizer = torch.optim.Adam(model.parameters(), lr=1e-3)
scheduler = WarmupHoldAnnealOneMinusSquareRoot(
    optimizer,
    warmup_ratio=0.01,
    hold_ratio=0.19,
    max_steps=10000,
    min_lr=1e-6
)
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
- [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

**Review by titu1994 — APPROVED** [review]
Thank you for your contribution ! The PR looks correct and has tests that pass so it's good to merge from my end

File: nemo/core/optim/lr_scheduler.py [hunk 1]
```diff
@@ -224,6 +224,24 @@ def get_lr(self):
 
         return self._get_lr(step)
 
+# REF: https://arxiv.org/html/2408.11029, with modification on enforcing min_lr
+class WarmupHoldAnnealOneMinusSquareRoot(WarmupHoldPolicy):
```
Comment (marcromeyn) [inline nemo/core/optim/lr_scheduler.py:228]: Would be great if a doc-strings could be added, like the other classes in this file. In the doc-string the paper could be referred as well.

**Comment by ericharper** [comment]
@hawkoli1987 , thanks for the contribution! Could you respond to the reviewer comments so we can merge this?

**Comment by hawkoli1987** [comment]
> @hawkoli1987 , thanks for the contribution! Could you respond to the reviewer comments so we can merge this?

Thanks @ericharper for your review! Sorry for the delay, I have updated the docstring and updated the PR. First time making PR to NeMo, please guide me along!

**Comment by ashors1** [comment]
@marcromeyn Could you give this PR a final review? @hawkoli1987 It also looks like there's a merge conflict. Could you please help to resolve that so we can get this PR in? Thank you both!

File: nemo/lightning/pytorch/optim/lr_scheduler.py [hunk 2]
```diff
@@ -484,3 +486,55 @@ def scheduler(self, model, optimizer):
             # Metric to to monitor for schedulers like `ReduceLROnPlateau`
             "monitor": self.monitor,
         }
+
+
+class WarmupHoldAnnealScheduler(LRSchedulerModule):
```
Comment (marcromeyn) [inline nemo/lightning/pytorch/optim/lr_scheduler.py:491]: Doc string missing here

**Comment by hawkoli1987** [comment]
conflict resolved, thanks!

File: nemo/lightning/pytorch/optim/lr_scheduler.py (hunk 2 above)
Comment (ashors1) [reply nemo/lightning/pytorch/optim/lr_scheduler.py:493]: @hawkoli1987 are you able to add a docstring here to address this comment?

Comment (verdimrc) [reply nemo/lightning/pytorch/optim/lr_scheduler.py:493]: docstring added.

**Review by ericharper — DISMISSED** [review]
LGTM. Thanks!
